nil pointer in exec after timeouts

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x5eac21]
goroutine 1393738 [running]:
github.com/gocql/gocql.(Conn).exec(0xc20813cb40, 0x7f4eeaf03110, 0xc2083803c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
/go/src/github.com/gocql/gocql/conn.go:393 +0x391
github.com/gocql/gocql.(Conn).executeQuery(0xc20813cb40, 0xc2083fc9c0, 0x9c76e0)
/go/src/github.com/gocql/gocql/conn.go:529 +0x5f4
github.com/gocql/gocql.(Session).executeQuery(0xc208038820, 0xc2083fc9c0, 0x7f4650)
/go/src/github.com/gocql/gocql/session.go:222 +0x1da
github.com/gocql/gocql.(Query).Iter(0xc2083fc9c0, 0xc2083fc9c0)
/go/src/github.com/gocql/gocql/session.go:613 +0x107
github.com/gocql/gocql.(*Query).Exec(0xc2083fc9c0, 0x0, 0x0)
/go/src/github.com/gocql/gocql/session.go:603 +0x3a
```

@xoraes yeah I decided to remove timeouts until I could reproduce this error and clearly understand the races going on.

I think I have been able to create a test that reproduces this and have a fix in place to enable timeouts again.

The failure shown is as follows,
```
=== RUN TestQueryTimeoutReuseStream-4
==================
WARNING: DATA RACE
Read by goroutine 9:
  github.com/gocql/gocql.(*Conn).recv()
      /Users/chris/dev/go/src/github.com/gocql/gocql/conn.go:328 +0x295
  github.com/gocql/gocql.(*Conn).serve()
      /Users/chris/dev/go/src/github.com/gocql/gocql/conn.go:303 +0x47

Previous write by goroutine 5:
  github.com/gocql/gocql.(*Conn).exec()
      /Users/chris/dev/go/src/github.com/gocql/gocql/conn.go:407 +0x82c
  github.com/gocql/gocql.(*Conn).executeQuery()
      /Users/chris/dev/go/src/github.com/gocql/gocql/conn.go:529 +0x930
  github.com/gocql/gocql.(*Session).executeQuery()
      /Users/chris/dev/go/src/github.com/gocql/gocql/session.go:222 +0x2b1
  github.com/gocql/gocql.(*Query).Iter()
      /Users/chris/dev/go/src/github.com/gocql/gocql/session.go:613 +0x176
  github.com/gocql/gocql.(*Query).Exec()
      /Users/chris/dev/go/src/github.com/gocql/gocql/session.go:603 +0x47
  github.com/gocql/gocql.TestQueryTimeoutReuseStream()
      /Users/chris/dev/go/src/github.com/gocql/gocql/conn_test.go:530 +0x399
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:447 +0x133

Goroutine 9 (running) created at:
  github.com/gocql/gocql.Connect()
      /Users/chris/dev/go/src/github.com/gocql/gocql/conn.go:185 +0xa31
  github.com/gocql/gocql.(*SimplePool).connect()
      /Users/chris/dev/go/src/github.com/gocql/gocql/connectionpool.go:228 +0x2ce
  github.com/gocql/gocql.NewSimplePool()
      /Users/chris/dev/go/src/github.com/gocql/gocql/connectionpool.go:205 +0x7b4
  github.com/gocql/gocql.NewSession()
      /Users/chris/dev/go/src/github.com/gocql/gocql/session.go:63 +0x174
  github.com/gocql/gocql.(*ClusterConfig).CreateSession()
      /Users/chris/dev/go/src/github.com/gocql/gocql/cluster.go:104 +0x86
  github.com/gocql/gocql.TestQueryTimeoutReuseStream()
      /Users/chris/dev/go/src/github.com/gocql/gocql/conn_test.go:522 +0x1b5
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:447 +0x133

Goroutine 5 (running) created at:
  testing.RunTests()
      /usr/local/go/src/testing/testing.go:555 +0xd4e
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:485 +0xe0
  main.main()
      github.com/gocql/gocql/_test/_testmain.go:292 +0x28c
==================
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x70 pc=0x84cea]
```

--- a/conn.go
+++ b/conn.go
@@ -348,6 +348,7 @@ func (c *Conn) recv() error {
 	select {
 	case call.resp <- err:
 	case <-call.timeout:
+		c.releaseStream(head.stream)
 	case <-c.quit:
 	}
 

--- a/conn.go
+++ b/conn.go
@@ -611,12 +611,11 @@ func (c *Conn) closeWithError(err error) {
 		req := &c.calls[id]
 		// we need to send the error to all waiting queries, put the state
 		// of this conn into not active so that it can not execute any queries.
+		atomic.StoreInt32(&req.waiting, 0)
 		select {
 		case req.resp <- err:
 		default:
 		}
-
-		close(req.resp)
 	}
 
 	c.conn.Close()

--- a/conn.go
+++ b/conn.go
@@ -376,6 +376,7 @@ func (c *Conn) handleTimeout() {
 
 func (c *Conn) exec(req frameWriter, tracer Tracer) (frame, error) {
 	// TODO: move tracer onto conn
+	start := time.Now()
 	stream := <-c.uniq
 
 	call := &c.calls[stream]
@@ -407,6 +408,7 @@ func (c *Conn) exec(req frameWriter, tracer Tracer) (frame, error) {
 		}
 	case <-time.After(c.timeout):
 		c.handleTimeout()
+		log.Printf("querytime=%v\n", time.Now().Sub(start))
 		return nil, ErrTimeoutNoResponse
 	}
 

--- a/conn.go
+++ b/conn.go
@@ -327,13 +327,15 @@ func (c *Conn) recv() error {
 	call := &c.calls[head.stream]
 	err = call.framer.readFrame(&head)
 	if err != nil {
-		return err
+		// only net errors should cause the connection to be closed. Though
+		// cassandra returning corrupt frames will be returned here as well.
+		if _, ok := err.(net.Error); ok {
+			return err
+		}
 	}
 
-	// once we get to here we know that the caller must be waiting and that there
-	// is no error.
 	select {
-	case call.resp <- nil:
+	case call.resp <- err:
 	default:
 		// in case the caller timedout
 	}

--- a/conn.go
+++ b/conn.go
@@ -113,6 +113,7 @@ type Conn struct {
 	started         bool
 
 	closed int32
+	quit   chan struct{}
 
 	timeouts int64
 }
@@ -171,6 +172,7 @@ func Connect(addr string, cfg ConnConfig, errorHandler ConnErrorHandler) (*Conn,
 		compressor:   cfg.Compressor,
 		auth:         cfg.Authenticator,
 		headerBuf:    make([]byte, headerSize),
+		quit:         make(chan struct{}),
 	}
 
 	if cfg.Keepalive > 0 {
@@ -343,8 +345,8 @@ func (c *Conn) recv() error {
 
 	select {
 	case call.resp <- err:
-	default:
-		c.releaseStream(head.stream)
+	case <-c.quit:
+		return nil
 	}
 
 	return nil
@@ -376,7 +378,6 @@ func (c *Conn) handleTimeout() {
 
 func (c *Conn) exec(req frameWriter, tracer Tracer) (frame, error) {
 	// TODO: move tracer onto conn
-	start := time.Now()
 	stream := <-c.uniq
 
 	call := &c.calls[stream]
@@ -388,7 +389,9 @@ func (c *Conn) exec(req frameWriter, tracer Tracer) (frame, error) {
 		framer.trace()
 	}
 
-	atomic.StoreInt32(&call.waiting, 1)
+	if !atomic.CompareAndSwapInt32(&call.waiting, 0, 1) {
+		return nil, errors.New("gocql: stream is busy or closed")
+	}
 	defer atomic.StoreInt32(&call.waiting, 0)
 
 	err := req.writeFrame(framer, stream)
@@ -408,8 +411,9 @@ func (c *Conn) exec(req frameWriter, tracer Tracer) (frame, error) {
 		}
 	case <-time.After(c.timeout):
 		c.handleTimeout()
-		log.Printf("querytime=%v\n", time.Now().Sub(start))
 		return nil, ErrTimeoutNoResponse
+	case <-c.quit:
+		return nil, ErrConnectionClosed
 	}
 
 	if v := framer.header.version.version(); v != c.version {
@@ -607,15 +611,13 @@ func (c *Conn) closeWithError(err error) {
 		return
 	}
 
+	close(c.quit)
+
 	for id := 0; id < len(c.calls); id++ {
 		req := &c.calls[id]
 		// we need to send the error to all waiting queries, put the state
 		// of this conn into not active so that it can not execute any queries.
-		atomic.StoreInt32(&req.waiting, 0)
-		select {
-		case req.resp <- err:
-		default:
-		}
+		atomic.StoreInt32(&req.waiting, -1)
 	}
 
 	c.conn.Close()
@@ -770,7 +772,8 @@ type inflightPrepare struct {
 }
 
 var (
-	ErrQueryArgLength    = errors.New("query argument length mismatch")
+	ErrQueryArgLength    = errors.New("gocql: query argument length mismatch")
 	ErrTimeoutNoResponse = errors.New("gocql: no response recieved from cassandra within timeout period")
 	ErrTooManyTimeouts   = errors.New("gocql: too many query timeouts on the connection")
+	ErrConnectionClosed  = errors.New("gocql: connection closed waiting for response")
 )

--- a/conn.go
+++ b/conn.go
@@ -337,6 +337,7 @@ func (c *Conn) recv() error {
 	select {
 	case call.resp <- err:
 	default:
+		c.releaseStream(head.stream)
 		// in case the caller timedout
 	}
 
@@ -365,7 +366,6 @@ func (c *Conn) handleTimeout() {
 func (c *Conn) exec(req frameWriter, tracer Tracer) (frame, error) {
 	// TODO: move tracer onto conn
 	stream := <-c.uniq
-	defer c.releaseStream(stream)
 
 	call := &c.calls[stream]
 	// resp is basically a waiting semaphore protecting the framer
@@ -382,6 +382,11 @@ func (c *Conn) exec(req frameWriter, tracer Tracer) (frame, error) {
 	}
 
 	err = <-call.resp
+
+	// dont release the stream if detect a timeout as another request can reuse
+	// that stream and get a response for the old request, which we have no
+	// easy way of detecting.
+	defer c.releaseStream(stream)
 
 	if err != nil {
 		return nil, err

--- a/conn.go
+++ b/conn.go
@@ -381,7 +381,12 @@ func (c *Conn) handleTimeout() {
 
 func (c *Conn) exec(req frameWriter, tracer Tracer) (frame, error) {
 	// TODO: move tracer onto conn
-	stream := <-c.uniq
+	var stream int
+	select {
+	case stream = <-c.uniq:
+	case <-c.quit:
+		return nil, ErrConnectionClosed
+	}
 
 	call := &c.calls[stream]
 	// resp is basically a waiting semaphore protecting the framer

--- a/conn.go
+++ b/conn.go
@@ -343,10 +343,12 @@ func (c *Conn) recv() error {
 		return nil
 	}
 
+	// we either, return a response to the caller, the caller timedout, or the
+	// connection has closed. Either way we should never block indefinatly here
 	select {
 	case call.resp <- err:
+	case <-call.timeout:
 	case <-c.quit:
-		return nil
 	}
 
 	return nil
@@ -357,6 +359,7 @@ type callReq struct {
 	resp    chan error
 	framer  *framer
 	waiting int32
+	timeout chan struct{} // indicates to recv() that a call has timedout
 }
 
 func (c *Conn) releaseStream(stream int) {
@@ -384,6 +387,7 @@ func (c *Conn) exec(req frameWriter, tracer Tracer) (frame, error) {
 	// resp is basically a waiting semaphore protecting the framer
 	framer := newFramer(c, c, c.compressor, c.version)
 	call.framer = framer
+	call.timeout = make(chan struct{})
 
 	if tracer != nil {
 		framer.trace()
@@ -410,6 +414,7 @@ func (c *Conn) exec(req frameWriter, tracer Tracer) (frame, error) {
 			return nil, err
 		}
 	case <-time.After(c.timeout):
+		close(call.timeout)
 		c.handleTimeout()
 		return nil, ErrTimeoutNoResponse
 	case <-c.quit:

--- a/conn.go
+++ b/conn.go
@@ -178,7 +178,7 @@ func Connect(addr string, cfg ConnConfig, errorHandler ConnErrorHandler) (*Conn,
 	}
 
 	for i := 0; i < cfg.NumStreams; i++ {
-		c.calls[i].resp = make(chan error, 1)
+		c.calls[i].resp = make(chan error)
 		c.uniq <- i
 	}
 
@@ -334,11 +334,17 @@ func (c *Conn) recv() error {
 		}
 	}
 
+	if !atomic.CompareAndSwapInt32(&call.waiting, 1, 0) {
+		// the waiting thread timed out and is no longer waiting, the stream has
+		// not yet been readded to the chan so it cant be used again,
+		c.releaseStream(head.stream)
+		return nil
+	}
+
 	select {
 	case call.resp <- err:
 	default:
 		c.releaseStream(head.stream)
-		// in case the caller timedout
 	}
 
 	return nil
@@ -346,11 +352,16 @@ func (c *Conn) recv() error {
 
 type callReq struct {
 	// could use a waitgroup but this allows us to do timeouts on the read/send
-	resp   chan error
-	framer *framer
+	resp    chan error
+	framer  *framer
+	waiting int32
 }
 
 func (c *Conn) releaseStream(stream int) {
+	call := &c.calls[stream]
+	framerPool.Put(call.framer)
+	call.framer = nil
+
 	select {
 	case c.uniq <- stream:
 	default:
@@ -376,20 +387,27 @@ func (c *Conn) exec(req frameWriter, tracer Tracer) (frame, error) {
 		framer.trace()
 	}
 
+	atomic.StoreInt32(&call.waiting, 1)
+	defer atomic.StoreInt32(&call.waiting, 0)
+
 	err := req.writeFrame(framer, stream)
 	if err != nil {
 		return nil, err
 	}
 
-	err = <-call.resp
+	select {
+	case err := <-call.resp:
+		// dont release the stream if detect a timeout as another request can reuse
+		// that stream and get a response for the old request, which we have no
+		// easy way of detecting.
+		defer c.releaseStream(stream)
 
-	// dont release the stream if detect a timeout as another request can reuse
-	// that stream and get a response for the old request, which we have no
-	// easy way of detecting.
-	defer c.releaseStream(stream)
-
-	if err != nil {
-		return nil, err
+		if err != nil {
+			return nil, err
+		}
+	case <-time.After(c.timeout):
+		c.handleTimeout()
+		return nil, ErrTimeoutNoResponse
 	}
 
 	if v := framer.header.version.version(); v != c.version {
@@ -404,9 +422,6 @@ func (c *Conn) exec(req frameWriter, tracer Tracer) (frame, error) {
 	if len(framer.traceID) > 0 {
 		tracer.Trace(framer.traceID)
 	}
-
-	framerPool.Put(framer)
-	call.framer = nil
 
 	return frame, nil
 }

--- a/conn_test.go
+++ b/conn_test.go
@@ -472,8 +472,7 @@ func TestPolicyConnPoolSSL(t *testing.T) {
 }
 
 func TestQueryTimeout(t *testing.T) {
-	t.Skip("skipping until query timeouts are enabled")
-	srv := NewTestServer(t, protoVersion2)
+	srv := NewTestServer(t, defaultProto)
 	defer srv.Stop()
 
 	cluster := NewCluster(srv.Address)
@@ -506,6 +505,31 @@ func TestQueryTimeout(t *testing.T) {
 	case <-time.After(10*time.Millisecond + db.cfg.Timeout):
 		// ensure that the query goroutines have been scheduled
 		t.Fatalf("query did not timeout after %v", db.cfg.Timeout)
+	}
+}
+
+func TestQueryTimeoutReuseStream(t *testing.T) {
+	srv := NewTestServer(t, defaultProto)
+	defer srv.Stop()
+
+	cluster := NewCluster(srv.Address)
+	// Set the timeout arbitrarily low so that the query hits the timeout in a
+	// timely manner.
+	cluster.Timeout = 1 * time.Millisecond
+	cluster.NumConns = 1
+	cluster.NumStreams = 1
+
+	db, err := cluster.CreateSession()
+	if err != nil {
+		t.Fatalf("NewCluster: %v", err)
+	}
+	defer db.Close()
+
+	db.Query("slow").Exec()
+
+	err = db.Query("void").Exec()
+	if err != nil {
+		t.Fatal(err)
 	}
 }
 
@@ -655,6 +679,18 @@ func (srv *TestServer) process(f *framer) {
 			f.writeInt(resultKindVoid)
 		case "timeout":
 			<-srv.quit
+			return
+		case "slow":
+			go func() {
+				f.writeHeader(0, opResult, head.stream)
+				f.writeInt(resultKindVoid)
+				f.wbuf[0] = srv.protocol | 0x80
+				select {
+				case <-srv.quit:
+				case <-time.After(50 * time.Millisecond):
+					f.finishWrite()
+				}
+			}()
 			return
 		default:
 			f.writeHeader(0, opResult, head.stream)

